### PR TITLE
Support SafeJoinPath and SafeJoinFilepath

### PR DESCRIPTION
**Pull request by wolfogre**
Follow #23371.

Introduce `SafeJoinPath` to replace `CleanPath`.

`CleanPath` could be misused as `CleanPath(path.Join("/tmp", "../etc/passwd"))`, and the result will be "/etc/passwd", when the right way is `path.Join("/tmp", CleanPath("../etc/passwd"))`.

However, it will be annoying if there are many paths like `path.Join("/tmp", CleanPath(input1), CleanPath(input2), CleanPath(input3))`. So `SafeJoinPath` could do it more simply: `SafeJoinPath("/tmp", input1, input2, input3)`.

If we really want something like `CleanPath`, we can just use `SafeJoinPath`: `CleanPath(p) == SafeJoinPath(p)`.

Also introduce `SafeJoinFilepath` for `filepath.Join`

### CODE REVIEW

File: modules/util/path.go [hunk 1]
```diff
@@ -248,3 +240,28 @@ func IsReadmeFileExtension(name string, ext ...string) (int, bool) {
...
+// SafeJoinFilepath is like filepath.Join, but it will prevent directory escaping.
+func SafeJoinFilepath(elem ...string) string {
+	separator := string(filepath.Separator)
+	elems := make([]string, len(elem))
+	for i, v := range elem {
+		elems[i] = filepath.Clean(separator + v)
+	}
+	if len(elem) > 0 && !strings.HasPrefix(elem[0], separator) {
+		return filepath.Join(elems...)[1:]
+	}
```
Comment (wxiaoguang) [inline modules/util/path.go]: For windows: `C:\foo` => `:\foo` ?

I do not think Gitea has Windows servers to run tests .... so the test code for Windows never runs?

Comment (wolfogre) [reply modules/util/path.go]: My bad, I thought the CI would run tests on Windows ...

I will fix it and run the tests on Windows to ensure it works.

**Comment by zeripath** [comment]
I don't think this is as helpful as you think it will be or as safe as you think it will be.

a) Do we really have that many places where we join multiple user provided paths? 
b) This doesn't prevent an initial "/". So you still always have to join the path with a base path.
c) There are times when user provided ".." are appropriate or joining ".." from within code is appropriate. This will likely prevent that.

Given that #23371 is IMO unsafe I think this needs more thought.

**Comment by wxiaoguang** [comment]
> I don't think this is as helpful as you think it will be or as safe as you think it will be ....

Although #23371 doesn't look good, I think we need some util functions for paths.

* There was incorrect `fs.Open(path.Clean(file))` , luckily it doesn't cause security problems.
* Since `action` is merged, Gitea needs to operate filesystems/stroages more and more, there are already many path-related security problems during review.


So, IMO:

* We need to clearly define every path usage, absolute or relative.
* Introduce some stable and clear path util functions.

**Comment by wolfogre** [comment]
Relaced by #23495

**Pull request closed without merging**